Switch geo validation to enum

MaineC opened the pull request:
Relates to #13608:

This switches handling coerce and ignore_malformed parameters to an enum to better reflect their relationship.

@cbuescher / @nknize / @colings86 can either of you have a look please whether those changes make sense and are in line with the issue above before I move on to the other geo queries.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java:
```diff
@@ -313,8 +291,8 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
-        builder.field("coerce", coerce);
-        builder.field("ignore_malformed", ignoreMalformed);
+        builder.field("coerce", validationMethod == GeoValidationMethod.COERCE);
+        builder.field("ignore_malformed", validationMethod == GeoValidationMethod.IGNORE_MALFORMED);
```
`ignoreMalformed()` would also be true for `GeoValidationMethod.COERCE`, I understand this doesn't change much in terms of the parser logic but maybe it would be more consistent to just use the getters here and rely on the logic they hide.

MaineC replied:
Good suggestion to use the getters. Will do.

colings86 replied:
I think we should change this so we output a `validation_method` field which can have the values `ignore_malformed`, `coerce` and `strict`. Then the parser should parse this as well as parsing the deprecated `coerce` and `ignore_malformed` boolean fields

cbuescher commented:
@MaineC Looks good, just left a few minor suggestions. Maybe @nknize can comment on whether the enum is clear enough from user perspective to remove the two independent setters from the java api.

colings86 commented on core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java:
```diff
@@ -164,22 +157,9 @@ public GeoBoundingBoxQueryBuilder setCornersOGC(String bottomLeft, String topRig
     public boolean coerce() {
-        return this.coerce;
+        return this.validationMethod == GeoValidationMethod.COERCE;
     }
```
I think we should remove this method and the ignoreMalformed() method and instead have a validationMethod() method here instead

MaineC commented:
@cbuescher addressed your comments. Will move on to other geo queries as well.

@colings86 Saw your comments just when I was about to write that I had addressed all comments. Fixed yours as well.

MaineC commented:
Also adapted GeoDistanceQueryBuilder, GeoDistanceRangeQueryBuilder, GeoPolygonQueryBuilder.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java:
```diff
@@ -88,7 +81,7 @@ public GeoBoundingBoxQueryBuilder(String fieldName) {
      */
     public GeoBoundingBoxQueryBuilder setCorners(double top, double left, double bottom, double right) {
-        if (!ignoreMalformed) {
+        if (! GeoValidationMethod.isIgnoreMalformed(validationMethod)) {
```
nit: I started to like the `(expression == false)` notation I saw in several places.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/GeoDistanceQueryParser.java:
```diff
@@ -103,7 +104,7 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                 } else if ("distance_type".equals(currentFieldName) || "distanceType".equals(currentFieldName)) {
                     geoDistance = GeoDistance.fromString(parser.text());
-                } else if (currentFieldName.endsWith(GeoPointFieldMapper.Names.LAT_SUFFIX)) {
+                } else if (currentFieldName.endsWith(GeoPointFieldMapper.Names.LAT_SUFFIX)) { // TODO IDE complains currentFieldName could be null here - add null check?
```
I saw this problem being dealt with in other place by setting currentFieldName to empty String. Worst that can happen then is that it is treated as fieldName in the query, which we should validate later and throw IAE then.

colings86 replied:
I don't think we should initialise to empty string. Why not add an assert here to check for null? This should never be null here as that would mean we got a value before a field name.

MaineC replied:
I like the idea of adding an assertion. The "this should never happen" thingy just bit me once too often in the past ;)

cbuescher commented:
@MaineC LGTM, left two minor suggestion but from my side this looks great.
